src/plonk/srs.rs

```diff
@@ -42,31 +43,155 @@ impl<C: CurveAffine> SRS<C> {
 
                 Ok(())
             }
+
+            fn copy(
+                &mut self,
+                permutation: usize,
+                left_wire: usize,
+                left_row: usize,
+                right_wire: usize,
+                right_row: usize,
+            ) -> Result<(), Error> {
+                let left: (usize, usize) = *self.copy[permutation]
+                    .get_mut(left_wire)
+                    .and_then(|wire| wire.get_mut(left_row))
+                    .ok_or(Error::BoundsFailure)?;
+
+                let right: (usize, usize) = *self.copy[permutation]
+                    .get_mut(right_wire)
+                    .and_then(|wire| wire.get_mut(right_row))
+                    .ok_or(Error::BoundsFailure)?;
+
+                if left == (left_wire, left_row) || right == (right_wire, right_row) {
+                    // Don't perform the copy constraint because it will undo
+                    // the effect of the permutation.
+                } else {
+                    *self.copy[permutation]
```

Comment:
We can just index into these vectors at this point since we've bounds checked these indicies previously.